include/wasm.hh

```diff
@@ -575,22 +630,32 @@ public:
 
 // Shared objects
 
-template<class T>
-class WASM_API_EXTERN Shared {
-public:
-  Shared() = delete;
-  ~Shared();
-  void operator delete(void*);
+template<class T> class WASM_API_EXTERN Shared;
+
+class Module;
+
+template<>
+class WASM_API_EXTERN Shared<Module> {
```

Comment:
FYI it's possible to remove `class Module;` here by instead writing:
```
template<>
class WASM_API_EXTERN Shared<class Module> {
```
I haven't done that because I expect most users of C++ to be surprised to discover that you're allowed to write a forward declaration using an elaborated type specifier inside the template argument of an explicit specialization. Regardless, it's an option that would be a little cleaner.

Reply:
`Shared<Module>` is used in the declaration of `Module`. So either we forward declare `Module` for `Shared<Module>`, or we forward-declare `Shared<Module>` for `Module`.

Reply:
Ah, right. Stylistically, it probably makes more sense to define `Shared<Module>` next to `Module`, but it's a minor point.

Reply:
Okay, I've done what I think you mean. If that isn't exactly what you meant, it's a quick PR to fix, I'd be happy to review it.

Reply:
Oh, I can't merge anyways. :)